minetest/minetest
# Add CSM 2D Drawing API

This is the cleaned up and bugfixed drawing portion of #10524.  Essentially, it adds a `Texture` class that allows drawing to custom textures and the screen, specifically pixels, filled and gradiented rectangles, textures, text, and items.  The `Texture` class has no support for in-game textures (that's for another PR).  In addition, it introduces a new versatile rectangle and positioning system which likely will be used in the actual formspec replacement API as well.

`client_lua_api.txt` is the best place to find the actual documentation

## To do

This PR is a Work in Progress.  There are a few small bugs, and fonts still need reworking.

Notably, drawing models is absent from this PR.  This is because I plan to add a good model system with support for multiple models in a scene (allowing e.g. the player model with armor on and a backpack), but that's too much for this PR, which is big enough already.  A way to draw lines, filled polygons, and other shapes will come in the 3D drawing PR as well.

## Discussion points

* Currently, positions are based on 0, not 1 like Lua, i.e. `{x=0, y=0}` is the top left, not `{x=1, y=1}`.  Should it be based on 1 to match Lua?

## How to test

This PR creates a new bundled CSM, `devtest_csm`, since there is no way to actually unit test this in the server-side `devtest`.  To test, enable this CSM and use the command `test_draw`.  The param `off` turns off the drawing, `window` turns on drawing to the Minetest window, `texture` draws it to a custom texture which is then drawn to the screen, `toggle_fill` turns filling the screen/texture with a color before drawing on and off, `dump_driver` prints the current video driver specs, and `test_ref` tests texture sizes via texture references.

For instance, here is what drawing to the window with filling on should look like:
![screenshot_20210112_175347](https://user-images.githubusercontent.com/31123645/104482966-6c86a080-557c-11eb-8ca7-12e460896a32.png)
